File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        
        int n = nums.size();
        vector<int> res(n, 1);

        // nums = [1, 2, 3, 4]
        // res = [1, 1, 1, 1]

        for(int i = 1; i < n; i++){
            res[i] = res[i - 1] * nums[i - 1];

            // res = [1, 1, 2, 6]
        }

        int postfix = 1;
        for(int i = n - 1; i >= 0; i--){

            // i = 3
            // res = [1, 1, 18, 6]
            // postfix = 1 -> 6 -> 
            res[i] *= postfix;
            postfix *= nums[i];
        }

        return res;
    }
};

File: Data Structures & Algorithms/top-k-elements-in-list/submission-1.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        
        // Step 1: Count frequencies
        HashMap<Integer, Integer> freqMap = new HashMap<>();
        for(int num : nums) {
            freqMap.put(num, freqMap.getOrDefault(num, 0) + 1);
        }
        
        // Step 2: Create buckets indexed by frequency
        List<Integer>[] buckets = new List[nums.length + 1];
        for(int i = 0; i < buckets.length; i++) {
            buckets[i] = new ArrayList<>();
        }
        
        // Step 3: Put elements into buckets based on their frequency
        for(int key : freqMap.keySet()) {
            int frequency = freqMap.get(key);
            buckets[frequency].add(key);
        }
        
        // Step 4: Collect top k elements from highest frequency buckets
        int[] result = new int[k];
        int index = 0;
        
        // Traverse from highest frequency to lowest
        for(int i = buckets.length - 1; i >= 0 && index < k; i--) {
            for(int num : buckets[i]) {
                result[index++] = num;
                if(index == k) break;
            }
        }
        
        return result;
    }
}